### PR TITLE
Improve pareto font figure tests for matplotlib

### PR DESCRIPTION
**Pull request by akawashiro**
## Motivation

This is `test_pareto_front` part of https://github.com/optuna/optuna/issues/2959.

## Description of the changes

Tests in this PR check values pareto front figure from matplotlib.

Note:
- I use `allclose_as_set` instead of `np.allclose` because the order of `figure.findobj(PathCollection)` may change.

### CODE REVIEW

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py [hunk 1]
```diff
@@ -1,14 +1,24 @@
...
 
 import optuna
 from optuna.visualization.matplotlib import plot_pareto_front
 
 
+def allclose_as_set(
+    points1: Union[List[List[float]], np.ndarray], points2: Union[List[List[float]], np.ndarray]
+) -> bool:
+    p1 = points1 if type(points1) == list else points1.tolist()  # type: ignore
+    p2 = points2 if type(points2) == list else points2.tolist()  # type: ignore
```
Comment (HideakiImamura) [inline tests/visualization_tests/matplotlib_tests/test_pareto_front.py:18]: Why do you need to ignore the mypy checking here? We should chech with mypy whenever we can.

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py [hunk 2]
```diff
@@ -72,6 +82,23 @@ def test_plot_pareto_front_2d(
...
+    path_offsets = list(map(lambda pc: pc.get_offsets(), figure.findobj(PathCollection)))
+    exists_pareto_front = any(
+        map(lambda po: allclose_as_set(po, pareto_front_points), path_offsets)
+    )
+    exists_dominated_trials = any(
+        map(lambda po: allclose_as_set(po, np.array([[1.0, 1.0]])), path_offsets)
+    )
+    assert exists_pareto_front
+    if include_dominated_trials:
+        assert exists_dominated_trials
```
Comment (HideakiImamura) [inline tests/visualization_tests/matplotlib_tests/test_pareto_front.py:100]: How about testing not only for `Test with `target_names` argument` but also `Test with three trials`?

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py [hunk 3]
```diff
@@ -153,6 +180,21 @@ def test_plot_pareto_front_3d(
...
+
+    path_offsets = list(map(lambda pc: pc.get_offsets(), figure.findobj(PathCollection)))
+    exists_pareto_front = any(
+        map(lambda po: allclose_as_set(po, pareto_front_points), path_offsets)
+    )
+    exists_dominated_trials = any(map(lambda po: allclose_as_set(po, [[1.0, 1.0]]), path_offsets))
+    assert exists_pareto_front
+    if include_dominated_trials:
+        assert exists_dominated_trials
+
```
Comment (HideakiImamura) [inline tests/visualization_tests/matplotlib_tests/test_pareto_front.py:197]: Ditto.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I have several comments. Could you take a look?

**Comment by HideakiImamura** [comment]
@himkt Could you review this PR if you have time?

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py (hunk 1 above)
Comment (akawashiro) [reply tests/visualization_tests/matplotlib_tests/test_pareto_front.py]: I found `isinstance` in `if` can handle `Uniont` type better and removed `# type: ignore`.
https://github.com/optuna/optuna/pull/3183/commits/08c77aac4141b0f3ccc0a2a9da51557a53cdabe5

File: tests/visualization_tests/matplotlib_tests/test_pareto_front.py (hunk 2 above)
Comment (akawashiro) [reply tests/visualization_tests/matplotlib_tests/test_pareto_front.py:117]: Test `three trials` also in  https://github.com/optuna/optuna/pull/3183/commits/aed169559609ea1d74c585aa8d4bca70d771cad6.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the fix. LGTM.

**Comment by akawashiro** [comment]
Thank you for your reviews.

**Review by himkt — APPROVED** [review]
@akawashiro I'm sorry for the late response and thank you so much for improving tests. LGTM.

**Pull request merged**